### DISCUSSION TITLE
Error Codes

### DISCUSSION THREAD

**Discussion by karlseguin** [duckdb/duckdb#11123] [Ideas]
It would be nice to expose more structured error data, specifically an error code. This would make detecting recoverable errors like a duplicate constraint error, or a table already existing, more reliable and flexible versus having to parse an error string.  

In the C API, this could either be exposed as a new function like `duckdb_result_error_code`  AND/OR the error string could, as part of the contract, always begin with a prefix (like "(code: 391) Catalog Error: Table with name "XYZ" already exists!".

**Comment by Mytherin** [comment]
`errors_as_json` provides this in some manner, although we don't provide detailed error information for all errors:

```sql
D set errors_as_json=true;
D select * from tbl;

{
   "exception_type":"Catalog",
   "exception_message":"Table with name tbl does not exist!\nDid you mean \"temp.information_schema.tables\"?",
   "name":"tbl",
   "candidates":"temp.information_schema.tables",
   "position":"14",
   "type":"Table",
   "error_subtype":"MISSING_ENTRY"
}
```

**Reply by karlseguin** [reply]
This help, thanks and for my use case, this is probably enough.

**Reply by kevinschaich** [reply]
Does this work in Python & other clients or only in the CLI? I'm having trouble getting it to work.
